llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp

```diff
@@ -1348,6 +1337,48 @@ void VPlanTransforms::addExplicitVectorLength(VPlan &Plan) {
   CanonicalIVIncrement->setOperand(0, CanonicalIVPHI);
   // TODO: support unroll factor > 1.
   Plan.setUF(1);
+
+  VPValue *BTC = Plan.getOrCreateBackedgeTakenCount();
+  auto IsHeaderMask = [BTC](VPValue *V) {
+    auto *CompareToReplace = dyn_cast<VPInstruction>(V);
+    return CompareToReplace &&
+           CompareToReplace->getOpcode() == Instruction::ICmp &&
+           CompareToReplace->getPredicate() == CmpInst::ICMP_ULE &&
+           CompareToReplace->getOperand(1) == BTC;
```

Comment:
> Must every ICMP_ULE of BTC be the header mask, regardless of its other operand?

At the moment yes, but I reworked this to walk users of header-masks.

The current placement of `addExplicitVectorLength` unfortunately means that `VPWidenCanonicalIVRecipe` may have been removed, if there's a matching `VPWidenIntOrFpInductionRecipe` (see TODO from the initial EVL PR; the current placement is to avoid introducing new uses of the canonical IV after introducing EVL)

So we also need to walk the users of `VPWidenIntOrFpInductionRecipe` that are canonical.

> Tail folding (an integral part of initial stripmining step according to roadmap) should arguably introduce a single abstract HeaderMask VPValue, retrievable from the loop region (as with its canonical IV), to be later materialized/legalized/lowered into concrete bump/widening/compare/active-lane-mask/no-mask-with-EVL recipes placed inside blocks?

Sounds good to me, pushing towards more gradual lowering (as follow-up)